File: frontends/Matlab/Preprocessing/elementpartitionhdg.m
function dmd = elementpartitionhdg(t,t2t,f,coupledinterface,nproc,metis)
 
[~,ne] = size(t);
dmd = cell(nproc,1);

if nproc==1
    i = 1;
    dmd{i}.elempart = 1:ne;
    dmd{i}.elempartpts = ne;
    dmd{i}.nbsd = [];
    dmd{i}.elem2cpu = [];
    dmd{i}.elemrecv = [];    
    dmd{i}.elemsend = [];
    dmd{i}.elemrecvpts = [];    
    dmd{i}.elemsendpts = [];
    if coupledinterface>0
      [~, inte] = find(f == coupledinterface);
      inte = unique(inte);
      dmd{i}.elempart = [setdiff((1:ne)', inte); inte];
      dmd{i}.elempartpts = [ne-length(inte) length(inte) 0];
      dmd{i}.intepartpts = [ne-length(inte) length(inte) 0 0];
    end    
    return;
end

elem2cpu = partition(t',ne,nproc,metis);

if coupledinterface>0
  [~, inte] = find(f == coupledinterface);
  inte = unique(inte);
end    

for i = 1:nproc
    dmd{i}.elem2rank = elem2cpu;
    disp(['element partition ' num2str(i)]); 
    intelem = find(elem2cpu == (i-1)); % elements in subdomain i    
    elem = neighboringelements(t2t, intelem); % all elements connected to elements in subdomain i         
    extelem = sort(setdiff(elem,intelem)); % exterior elements    
                                
    elem = neighboringelements(t2t, extelem); % all elements connected to exterior elements
    bndelem = intersect(elem,intelem);  % interface elements in subdmain i
    
    % split bndelem into bndelem1 and bndelem2
    if coupledinterface>0
      bndelem1 = intersect(bndelem, inte);
      bndelem2 = setdiff(bndelem, bndelem1);
      bndelem = [bndelem1; bndelem2];
    end
    
    dmd{i}.elempart = [setdiff(intelem,bndelem); bndelem; extelem]; % partitions of elements
    dmd{i}.elempartpts = [length(intelem)-length(bndelem) length(bndelem) length(extelem)];    
    if coupledinterface>0
      dmd{i}.intepartpts = [length(intelem)-length(bndelem) length(bndelem1) length(bndelem2) length(extelem)];    
    end
    dmd{i}.elem2cpu = elem2cpu(dmd{i}.elempart);
    nelem = dmd{i}.elempartpts;

    recvelem = extelem; % elements received from neighbors
    ind = xiny(recvelem(:), dmd{i}.elempart);    
    dmd{i}.elemrecv = [dmd{i}.elem2cpu(ind)+1 (sum(nelem(1:2))+1:sum(nelem))' recvelem];            
    [~,ind] = sortrows(dmd{i}.elemrecv); 
    %[~,ind] = sort(dmd{i}.elemrecv(:,1)); 
    dmd{i}.elemrecv = dmd{i}.elemrecv(ind,:);
    dmd{i}.nbsd = unique(dmd{i}.elemrecv(:,1)); % neighboring subdomains        
end

% store elements sent to neighboring subdomains to assemble the linear system
for k = 1:nproc    
    dmd{k}.elemsend = [];
end
for i = 1:nproc            
    for j = 1:length(dmd{i}.nbsd) % for each neighbor of cpu i
        % cpu k sends information to cpu i
        k = dmd{i}.nbsd(j); % neighbor k of cpu i
        ii = dmd{i}.elemrecv(:,1)==k; % indices 
        tm = dmd{i}.elemrecv(ii,:);   % only get the elements received from cpu k
        tm(:,1) = i;                  % change k to i, so that cpu k sennds elements to cpu i 
        tm(:,2) = xiny(tm(:,3),dmd{k}.elempart(:)); % local indices of elements sent to cpu i 
        dmd{k}.elemsend = [dmd{k}.elemsend; tm];        
    end    
end

for i = 1:nproc    
    for j = 1:length(dmd{i}.nbsd)
        dmd{i}.elemsendpts(j) = length(find(dmd{i}.elemsend(:,1)==dmd{i}.nbsd(j)));
        dmd{i}.elemrecvpts(j) = length(find(dmd{i}.elemrecv(:,1)==dmd{i}.nbsd(j)));
    end
    dmd{i}.nbsend = dmd{i}.elemsend(:,1);
    dmd{i}.nbrecv = dmd{i}.elemrecv(:,1);        
    dmd{i}.elemsend = dmd{i}.elemsend(:,2);
    dmd{i}.elemrecv = dmd{i}.elemrecv(:,2);        
end

%save dmdelem.mat dmd t2t elem2cpu;


% for i = 1:nproc  
%   [length(dmd{i}.elemsend) length(dmd{i}.elemrecv)]
%   [length(unique(dmd{i}.elemsend)) length(unique(dmd{i}.elemrecv))]
% end

function nbelem = neighboringelements(t2t, elem)

t2te = t2t(:,elem);            
nbelem = unique(t2te(:));
if nbelem(1) == 0
    nbelem(1) = [];
end

File: frontends/Python/Preprocessing/elementpartitionhdg.py
import numpy as np
from ainb import ainb
from sortrows import sortrows
from partition import partition
from neighboringelements import neighboringelements

def elementpartitionhdg(dmd, t, t2t, nproc, metis):
    nve, ne = t.shape

    if nproc == 1:
        i = 0
        dmd[i]['elempart'] = np.arange(0,ne).astype(int);
        dmd[i]['elempartpts'] = np.array([ne]).astype(int);
        dmd[i]['nbsd'] = [];
        #dmd[i].elem2cpu = reshape([0],1,1);
        dmd[i]['elemrecv'] = [];
        dmd[i]['elemsend'] = [];
        dmd[i]['elemrecvpts'] = [];
        dmd[i]['elemsendpts'] = [];
        # #dmd[i]['elempart'] = np.reshape(np.arange(ne), (1, ne))
        # #dmd[i]['elempartpts'] = np.reshape([ne], (1, 1))
        # dmd[i]['nbsd'] = np.reshape([0], (1, 1))
        # # dmd[i]['elem2cpu'] = np.reshape([0], (1, 1))
        # dmd[i]['elemrecv'] = np.reshape([0], (1, 1))
        # dmd[i]['elemsend'] = np.reshape([0], (1, 1))
        # dmd[i]['elemrecvpts'] = np.reshape([0], (1, 1))
        # dmd[i]['elemsendpts'] = np.reshape([0], (1, 1))
        return dmd

    elem2cpu = partition(t+1,ne,nproc,metis)[0]
    elem2cpu = np.array(elem2cpu).flatten().astype(int);

    for i in range(nproc):
        intelem = np.where(elem2cpu == i)[0]  # elements in subdomain i
        intelem = np.array(intelem).astype(int).flatten();
        elem = neighboringelements(t2t, intelem)  # all elements connected to elements in subdomain i
        extelem = np.sort(np.setdiff1d(elem, intelem))  # exterior elements

        elem = neighboringelements(t2t, extelem)  # all elements connected to exterior elements
        bndelem = np.intersect1d(elem, intelem)  # boundary elements in subdomain i

        tmp = np.concatenate((np.setdiff1d(intelem, bndelem), bndelem, extelem))
        dmd[i]['elempart'] = tmp.flatten('F'); # partitions of elements        
        dmd[i]['elempartpts'] = np.array([len(intelem)-len(bndelem), len(bndelem), len(extelem)]).flatten();
        # dmd[i]['elem2cpu'] = elem2cpu[dmd[i]['elempart']]
        nelem = dmd[i]['elempartpts']

        recvelem = extelem
        tmp = np.array(np.arange(np.sum(nelem[0:2])+1,np.sum(nelem)+1)).flatten().astype(int);
        elemrecv = np.zeros((len(recvelem),3)).astype(int);
        elemrecv[:,0] = elem2cpu[recvelem];
        elemrecv[:,1] = tmp-1;
        elemrecv[:,2] = recvelem;
        tmp,ind = sortrows(elemrecv)[0:2];
        dmd[i]['elemrecv'] = elemrecv[ind.flatten('F'),:];
        dmd[i]['nbsd'] = (np.sort(np.unique(dmd[i]['elemrecv'][:,0]))).astype(int); # neighboring subdomains

    # store elements sent to neighboring subdomains to assemble the linear system
    for k in range(0,nproc):
        dmd[k]['elemsend'] = np.zeros(0);

    for i in range(0,nproc):
        for j in range(0,np.size(dmd[i]['nbsd'])):
            # cpu k sends information to cpu i
            k = dmd[i]['nbsd'][j];
            ii = np.nonzero(dmd[i]['elemrecv'][:,0] == k)[0];
            tm = dmd[i]['elemrecv'][ii,:];
            tm[:,0] = i;
            tm[:,1] = ainb(tm[:,2].flatten(), dmd[k]['elempart'].flatten());
            if np.size(dmd[k]['elemsend'])==0:
                dmd[k]['elemsend'] = tm;
            else:
                dmd[k]['elemsend'] = np.vstack((dmd[k]['elemsend'], tm));

    for i in range(0,nproc):
        dmd[i]['elemsendpts'] = np.zeros(np.size(dmd[i]['nbsd'])).astype(int);
        dmd[i]['elemrecvpts'] = np.zeros(np.size(dmd[i]['nbsd'])).astype(int);
        for j in range(0,np.size(dmd[i]['nbsd'])):
            dmd[i]['elemsendpts'][j] = np.size(np.nonzero(dmd[i]['elemsend'][:,0] == dmd[i]['nbsd'][j])[0]);
            dmd[i]['elemrecvpts'][j] = np.size(np.nonzero(dmd[i]['elemrecv'][:,0] == dmd[i]['nbsd'][j])[0]);

        dmd[i]['elemsend'] = dmd[i]['elemsend'][:,1];
        dmd[i]['elemrecv'] = dmd[i]['elemrecv'][:,1];

    return dmd  # elem2cpu; #elempart, elempartall

# Example usage:
# dmd = some_initial_value
# t, t2t, nproc, metis = some_other_values
# dmd = elementpartitionhdg(dmd, t, t2t, nproc, metis)
